---

layout: page
title: "Publications"
permalink: /publications
nav: true
order: 3

---

<head>

<style>
img {
border: 0px solid #787878;
    max-width: 200px;
    max-height: 130px;
}
#txt {
    font-size:11pt;
}

</style>
</head>

<h5>* = equal contribution</h5>
<h5>' = co-corresponding</h5>

<table>
  <tr>
    <td><img src="/assets/pst.png" id="fig"></td>
    <td><b>Endowing protein language models with structural knowledge</b><br>
	Dexiong Chen*, Philip Hartout*, Paolo Pellizzoni, Carlos Oliver, and Karsten Borgwardt <br>
	<i>arXiV (2024) </i> (<a href="https://arxiv.org/pdf/2401.14819.pdf">preprint</a>) </td>
  </tr>

  <tr>
    <td><img src="/assets/gmsm.png" id="fig"></td>
    <td><b>Graph-matching-based learning of substitution matrices for biological structures with functional priors</b><br>
	Paolo Pellizzoni*, Carlos Oliver*, Karsten Borgwardt <br>
	<i>RECOMB (2024) </i> (preprint coming soon.) </td>
  </tr>
  <tr>
    <td><img src="/assets/rnamigos2.png" id="fig"></td>
    <td><b>Semi-supervised learning and large-scale docking data accelerate RNA virtual screening</b><br>
	Juan G. Carvajal Pati&ntilde;o*, Vincent Mallet*, David Becerra, L. Fernando Ni&ntilde;o V, Carlos Oliver', Jerome Waldispuhl'<br>
	<i>bioRxiv (2023) </i> (<a href="https://www.biorxiv.org/content/10.1101/2023.11.23.568394v1.full.pdf+html">preprint</a>) </td>
  </tr>
  <tr>
    <td><img src="/assets/amr.png" id="fig"></td>
    <td><b>Multimodal learning in clinical proteomics: enhancing antimicrobial resistance models with chemical information</b><br>
	Giovanni Visonà*, Diane Duroux*, Lucas Miranda, Emese Sükei, Yiran Li, Karsten Borgwardt', and Carlos Oliver' <br>
	<i>Bioinformatics (2023) </i>  (<a href="https://academic.oup.com/bioinformatics/article/39/12/btad717/7450077">paper</a>) </td>
  </tr>

  <tr>
    <td><img src="/assets/ps_logo.png" id="fig"></td>
    <td><b>ProteinShake: Building Datasets and Benchmarks for Deep Learning on Protein Structures</b> <br>
	Tim Kucera*, Carlos Oliver*, Dexiong Chen, Karsten Borgwardt <br>
	<i>NeurIPS Datasets and Benchmarks (2023) </i> (<a href="https://openreview.net/pdf?id=27vPcG4vKV">preprint</a>)</td>
  </tr>
  <tr>
    <td><img src="/assets/mds.png" id="fig"></td>
    <td><b>Unsupervised Manifold Alignment with Joint Multidimensional Scaling</b> <br>
	Dexiong Chen, Bowen Fan, Carlos Oliver, Karsten Borgwardt <br>
	<i>ICLR 2022</i>(<a href="https://arxiv.org/abs/2207.02968">preprint</a>)</td>
  </tr>
  <tr>
    <td><img src="/assets/motifiesta.png" id="fig"></td>
    <td><b>Approximate Network Motif Mining via Graph Learning </b> <br>
	Carlos Oliver, Dexiong Chen, Vincent Mallet, Pericles Philippopoulos, Karsten Borgwardt <br>
	<i>ArXiV (2022)</i> (<a href="https://arxiv.org/abs/2206.01008">preprint</a>)</td>
  </tr>
  <tr>
    <td><img src="/assets/rgl_fig.png" id="fig"></td>
    <td><b>RNAGlib: A Python Package for RNA 2.5D Graphs. </b> <br>
	Vincent Mallet, Carlos Oliver, Jonathan Broadbent,  William L. Hamilton, Jerome Waldispuhl<br>
	<i> Bioinformatics (2021) </i>(<a href="https://arxiv.org/abs/2109.04434">preprint</a> | <a href="https://academic.oup.com/bioinformatics/article/38/5/1458/6462185?login=true">article</a>)</td>
  </tr>
  <tr>
    <td><img src="/assets/relattentional.png" id="fig"></td>
    <td><b>Edge Similarity-Aware Graph Neural Networks. </b> <br>
	Vincent Mallet, Carlos Oliver, William L. Hamilton<br>
	<i> arXiv (2021)</i>(<a href="https://arxiv.org/abs/2109.09432">preprint</a>)</td>
  </tr>
  <tr>
    <td><img src="/assets/vernal.png" id="fig"></td>
    <td><b>VeRNAl: A tool for fuzzy network motif mining in RNA. </b> <br>
	Carlos Oliver, Vincent Mallet, Pericles Philippopoulos, William L. Hamilton, Jerome Waldispuhl<br>
	<i> RECOMB 2021 & Bioinformatics </i>(<a href="https://arxiv.org/abs/2009.00664">preprint</a> | <a href="https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btab768/6428528?redirectedFrom=fulltext">article</a>)</td>
 </tr>
  <tr>
    <td><img src="/assets/optimol.png" id="fig"></td>
    <td><b>OptiMol: Optimization of binding affinities in chemical space for drug discovery. </b> <br>
	Jacques Boitreaud, Vincent Mallet, Carlos Oliver, Jerome Waldispuhl<br>
	<i> ACM JCIM (2020) </i>(<a href="https://pubs.acs.org/doi/10.1021/acs.jcim.0c00833">article</a> | <a href="https://www.biorxiv.org/content/10.1101/2020.05.23.112201v2">preprint</a>)</td>
  </tr> 

  <tr>
    <td><img src="/assets/rnamigos.png" id="fig"></td>
    <td><b>Augmented RNA base pairing networks imprint small molecule binding preferences. </b><br>
	Carlos Oliver, Vincent Mallet, Roman Sarrazin Gendron, Vladimir Reinharz, William L. Hamilton, Nicolas Moitessier, Jerome Waldispuhl <br>
	<i>Nucleic Acids Research (2020) </i> (<a href="https://academic.oup.com/nar/article/doi/10.1093/nar/gkaa583/5870337">article</a>|<a href="https://www.biorxiv.org/content/10.1101/701326v3">preprint</a>)</td>
  </tr>

  <tr>
    <td><img src="/assets/dips.png" id="fig"></td>
    <td><b>Difficulty Scaling in Proof of Work for Decentralized Problem Solving.</b><br>
	   Pericles Philippopoulos, Alessandro Ricottone, Carlos Oliver <br>
	   <i> Ledger Journal (2020) </i> (<a href="http://ledger.pitt.edu/ojs/ledger/article/view/194">article</a>)</td>
  </tr>

  <tr>
    <td><img src="/assets/bp2.png"></td>
    <td><b>Stochastic Sampling of Structural Contexts Improves the Scalability and Accuracy of RNA 3D Module Identification</b><br>
	   Roman Sarrazin Gendron, Hua-Ting Yao, Vladimir Reinharz, Carlos Oliver, Yann Pony, Jerome Waldispuhl<br>
	   <i> Accepted at RECOMB 2020</i> (<a href="https://www.biorxiv.org/content/biorxiv/early/2020/01/18/834762.full.pdf">preprint</a>) </td>
  </tr>

  <tr>
    <td><img src="/assets/tarlig.png"></td>
    <td><b>Leveraging binding site structure for drug discovery with point-cloud methods</b><br>
	   Vincent Mallet, Carlos Oliver, Nicolas Moitessier, Jerome Waldispuhl <br>
	   <i> ArXiv (2019) </i>(<a href="https://arxiv.org/pdf/1905.12033.pdf">preprint</a>) </td>
  </tr>

  <tr>
    <td><img src="/assets/bp1.png"></td>
    <td><b>Automated, customizable and efficient identification of 3D base pair modules with BayesPairing</b><br>
	   Roman Sarrazin Gendron, Vladimir Reinharz, Carlos Oliver, Nicolas Moitessier, Jerome Waldispuhl <br>
	   <i> Nucleic Acids Research (2019) </i>(<a href="https://academic.oup.com/nar/article/47/7/3321/5369007?searchresult=1">article</a>) </td>
   </tr>

  <tr>
    <td><img src="/assets/maternal.png"></td>
    <td><b>On the emergence of structural complexity in RNA replicators. </b><br>
	   Carlos Oliver, Vladimir Reinharz, Jerome Waldispuhl  <br>
	   <i> RNA (2018)  </i>(<a href="https://rnajournal.cshlp.org/content/25/12/1579.short">article</a>)</td>
   </tr>

  <tr>
    <td><img src="/assets/crick.png"></td>
    <td><b>Proposal for a fully decentralized blockchain and proof-of-work algorithm for solving NP-complete problems</b><br>
	   Carlos Oliver, Alessandro Ricottone, Pericles Philippopoulos <br>
	   <i> ArXiv (2017)</i>(<a href="https://arxiv.org/abs/1708.09419">preprint</a>)</td>
   </tr>

  <tr>
    <td><img src="/assets/tub.png"></td>
    <td><b>Concerted millisecond timescale dynamics in the intrinsically disordered carboxyl terminus of gamma‐tubulin induced by mutation of a conserved tyrosine residue</b><br>
	   Jason Harris, Maria Shadrina, Carlos Oliver, Anthony Mittermaier, Jackie Vogel <br>
	   <i> Protein Science (2018) </i>(<a href="https://onlinelibrary.wiley.com/doi/full/10.1002/pro.3345">article</a>)</td>
   </tr>

  <tr>
    <td><img src="/assets/cisnat.png"></td>
    <td><b>The distribution and evolution of Arabidopsis thaliana cis natural antisense transcripts</b><br>
	   Jonathan Bouchard, Carlos Oliver, Paul Harrison <br>
	   <i> BMC Genomics (2015) </i>(<a href="https://link.springer.com/article/10.1186/s12864-015-1587-0">article</a>)</td>
   </tr>
</table>

<!--Other:-->

<!--* Ph.D. Comprehensive Exam Literature Review ([PDF]({{ site.url  }}/assets/review_cgo.pdf))-->
<!--* M.Sc. Thesis ([PDF]({{ site.url   }}/assets/msc_thesis.pdf))-->
